File: src/test/python/reduce/reduce_extract_session_counts.py
#!/usr/bin/env python

from operator import itemgetter
from itertools import groupby
import sys

def read_mapper_output(file, separator='\t'):
    for line in file:
        yield line.rstrip().split(separator, 1)

def main(separator='\t'):
    # input comes from STDIN (standard input)
    data = read_mapper_output(sys.stdin, separator=separator)
    
    # groupby groups multiple user_id sessions pairs by user_id,
    # and creates an iterator that returns consecutive keys and their group:
    #   current_key - string containing session_id
    #   group - iterator yielding all items
    for current_key, group in groupby(data, itemgetter(0)):
        try:
            # Setup the session list
            sessionList = [0] * 19
            for k, v in group:
                            
                if 'Account' in v:
                    sessionList[0] += 1
                    
                if 'AddToQueue' in v:
                    sessionList[1] += 1
                    
                if 'Advance' in v:
                    sessionList[2] += 1
                
                if 'Home' in v:
                    sessionList[3] += 1
                    
                if 'Hover' in v:
                    sessionList[4] += 1
                    
                if 'ItemPage' in v:
                    sessionList[5] += 1
                    
                if 'Login' in v:
                    sessionList[6] += 1
                    
                if 'Logout' in v:
                    sessionList[7] += 1
                    
                if 'Pause' in v:
                    sessionList[8] += 1
                    
                if 'Play' in v:
                    sessionList[9] += 1
                    
                if 'Position' in v:
                    sessionList[10] += 1
                    
                if 'Queue' in v:
                    sessionList[11] += 1
                
                if 'Rate' in v:
                    sessionList[12] += 1
                    
                if 'Recommendations' in v:
                    sessionList[13] += 1
                    
                if 'Resume' in v:
                    sessionList[14] += 1
                    
                if 'Search' in v:
                    sessionList[15] += 1
                    
                if 'Stop' in v:
                    sessionList[16] += 1
                
                if 'VerifyPassword' in v:
                    sessionList[17] += 1
                
                if 'WriteReview' in v:
                    sessionList[18] += 1
                
            print '%s%s%s' % (current_key, "\t", "\t".join(map(str, sessionList)))
                            
        except ValueError:
            # count was not a number, so silently discard this item
            pass

if __name__ == "__main__":
    main()

File: src/main/java/ccp/shorsman/task2/ReduceExtractSessionCounts.java
// ReduceExtractSessionCounts creates a csv for input into the Cloudera ML jobs.  Each column
// represents a duration for each action during a session.  The last two columns are a count of 
// how many unique items were watched during the session (using java.util.Set). 
package ccp.shorsman.task2;

import java.io.IOException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Reducer;
import org.joda.time.DateTime;
import org.joda.time.Seconds;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

public class ReduceExtractSessionCounts extends Reducer<Text, Text, Text, Text> {

	@Override
	public void reduce(Text key, Iterable<Text> values, Context context)
			throws IOException, InterruptedException {

		int[] sessionArray;
		
		Set<String> uniqueMovies = new HashSet<String>();
		Set<String> uniqueTvEpisodes = new HashSet<String>();

		sessionArray = new int[16];

		// Go through all values to and group together like actions
		for (Text value : values) {
			// Split the value on between action and date
			String[] valuesList = value.toString().split(",");
			String action = valuesList[0];
			String createdAt = valuesList[1];
			int actionDuration = Integer.parseInt(valuesList[2]);
			String itemId = valuesList[3];

			// For clustering we'll group together some events based on
			// likeness:
			// Login & Logout
			if (action.toString().equalsIgnoreCase("Login"))
				sessionArray[0] += actionDuration;
			if (action.toString().equalsIgnoreCase("Logout"))
				sessionArray[1] += actionDuration;

			// Account & VerifyPassword (group)
			if (action.toString().equalsIgnoreCase("Account"))
				sessionArray[2] += actionDuration;
			if (action.toString().equalsIgnoreCase("VerifyPassword"))
				sessionArray[2] += actionDuration;

			// WriteReview & Rate (group)
			if (action.toString().equalsIgnoreCase("Rate"))
				sessionArray[3] += actionDuration;
			if (action.toString().equalsIgnoreCase("WriteReview"))
				sessionArray[3] += actionDuration;

			// Queue, AddToQueue, ItemPage must be when a user adds a movie to
			// be watched (group)
			if (action.toString().equalsIgnoreCase("AddToQueue"))
				sessionArray[4] += actionDuration;
			if (action.toString().equalsIgnoreCase("Queue"))
				sessionArray[4] += actionDuration;
			if (action.toString().equalsIgnoreCase("ItemPage"))
				sessionArray[4] += actionDuration;

			// Advance, Position and Resume (all have a marker so group)
			// Are there lots of session which resume in the middle of a film?
			if (action.toString().equalsIgnoreCase("Advance"))
				sessionArray[5] += actionDuration;
			if (action.toString().equalsIgnoreCase("Position"))
				sessionArray[5] += actionDuration;
			if (action.toString().equalsIgnoreCase("Resume"))
				sessionArray[5] += actionDuration;

			// Home page
			if (action.toString().equalsIgnoreCase("Home"))
				sessionArray[6] += actionDuration;

			// Hover action
			if (action.toString().equalsIgnoreCase("Hover"))
				sessionArray[7] += actionDuration;

			// Stop & Pause represent some sort of interruption
			if (action.toString().equalsIgnoreCase("Pause"))
				sessionArray[8] += actionDuration;
			if (action.toString().equalsIgnoreCase("Stop"))
				sessionArray[9] += actionDuration;

			// Recommendations
			if (action.toString().equalsIgnoreCase("Recommendations"))
				sessionArray[10] += actionDuration;

			// Search
			if (action.toString().equalsIgnoreCase("Search"))
				sessionArray[11] += actionDuration;
			
			// Split Play between Movies and TV
			// Add a count of how many movies and TV was watched
			if (action.toString().equalsIgnoreCase("PlayTV")) {
				sessionArray[12] += actionDuration;
				
				uniqueTvEpisodes.add(itemId);
			}
				
			
			if (action.toString().equalsIgnoreCase("PlayMovie")) {
				sessionArray[14] += actionDuration;
				
				uniqueMovies.add(itemId);
			}

		}
		
		sessionArray[13] = uniqueTvEpisodes.size();
		sessionArray[15] = uniqueMovies.size();
		
		StringBuffer newValue = new StringBuffer("");
		for (int i = 0; i < sessionArray.length; i++) {
			if (i > 0)
				newValue.append(",");

			newValue.append(sessionArray[i]);

		}
		context.write(key, new Text(newValue.toString()));
	}
}
